linkerd/connection-header/src/lib.rs

```
@@ -0,0 +1,264 @@
+        while buf.len() < PREFACE_LEN {
+            if io.read_buf(buf).await? == 0 {
+                return Ok(None);
+            }
+        }
+
+        // Advance the buffer past the preface if it matches.
+        if &buf.bytes()[..PREFACE.len()] != PREFACE {
+            return Ok(None);
+        }
+        buf.advance(PREFACE.len());
+
+        // Read the message length. If it is larger than our allowed buffer
+        // capacity, fail the connection.
+        let msg_len = buf.get_u32() as usize;
+        if msg_len > buf.capacity() + PREFACE_LEN {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Message length exceeds capacity",
+            ));
+        }
+
+        // Free up parsed preface data and ensure there's enough capacity for
+        // the message.
+        buf.reserve(msg_len);
+        while buf.len() < msg_len {
+            if io.read_buf(buf).await? == 0 {
+                return Err(io::Error::new(
+                    io::ErrorKind::UnexpectedEof,
+                    "Full header message not provided",
+                ));
+            }
+        }
+
+        // Take the bytes needed to parse the message and leave the remaining
+        // bytes in the caller-provided buffer.
+        let msg = buf.split_to(msg_len);
+        Self::decode(msg.freeze())
+    }
+
+    // Decodes a protobuf message from the buffer.
+    #[inline]
+    fn decode<B: Buf>(buf: B) -> io::Result<Option<Self>> {
+        let h = proto::Header::decode(buf)
+            .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "Invalid header message"))?;
+
+        let name = if h.name.len() == 0 {
+            None
+        } else {
+            let n = Name::from_str(&h.name)
+                .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "Invalid name"))?;
+            Some(n)
+        };
+
+        Ok(Some(Self {
+            name,
+            port: h.port as u16,
```

this will silently truncate the port if it's greater than 65535, should we return an error instead?

if we got a port number greater than `u16::max` and we know the remote peer is using the same code (which will only encode `Header` structs with `u16` port values), seeing a value higher than `u16::max` probably indicates that the message is corrupted or we decoded it wrongly, right?